Pack optimizations

tzulberti-jampp opened the pull request.

klaussfreire committed:
Various sequence packing optimizations

klaussfreire committed:
Use unpack_from

klaussfreire committed:
List/array packing optimizations

- Automatic and efficient nparray -> list casting
- Changed format of nparray dtype to avoid wrapping overhead
  both in size and performance for common dtypes

klaussfreire committed:
Optimize tuple number range checks

klaussfreire committed:
Integers as equality keys

klaussfreire committed:
Merge branch 'object_pool' into pack_optimizations

klaussfreire committed:
Merge branch 'object_pool' into pack_optimizations

klaussfreire committed:
Wide frozensets and some frozenset fixes and tests

klaussfreire committed:
Merge branch 'object_pool' into pack_optimizations

klaussfreire committed:
Optimize object packing

Reduce the number of attribute reads by 3. Earlier, we were doing:
- hasattr then getattr in get_bitmaps
- getattr in get_packable

Now, get_bitmaps just does has_attr and captures AttributeError,
then passes the values back to get_packable

klaussfreire committed:
Declare values as list

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -332,12 +361,18 @@ class WRAPPED:
     pass
 
 
+WRAP_MASK = cython.declare(object, 1L << 64)
```
Are this constants defined here for any particular reason?

klaussfreire replied:
To name them

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -332,12 +361,18 @@ class WRAPPED:
     if isinstance(xid, (int, long)):
         # We keep numeric ids numeric
-        xid |= 0xFL << 64
+        xid |= WRAP_MASK
```
Shouldn't this be FLAG_MASK? or there was a bug about that?

klaussfreire replied:
Nope. FLAG_MASK is a mask to get all flags, WRAP_MASK is the "wrapped" flag only

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -441,72 +511,132 @@ def unpack_from(cls, buf, offs, idmap = None):
+            all_int = all_intlong = all_float = 1
+            minval = maxval = 0
+            for x in obj:
+                if all_int and type(x) is not int:
```
Is this faster than checking `isinstance(x, int)`?

klaussfreire replied:
Yes, considerably. Also, not exactly the same. This doesn't accept subclasses, and in many cases we don't want to accept subclasses.

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -441,72 +511,132 @@ def unpack_from(cls, buf, offs, idmap = None):
+            all_int = all_intlong = all_float = 1
+            minval = maxval = 0
+            for x in obj:
+                if all_int and type(x) is not int:
```
Why don't just calculate the type once: `x_type = type(x)`?

klaussfreire replied:
It would be counterproductive. Check the generated code.

klaussfreire committed:
Fix map_file to return a proper slice of the buffer

This matters when page alignment forces the mapping to
be bigger, and the user depends on the buffer bounds
to be exact

klaussfreire committed:
Minor proxied_list optimizations

klaussfreire committed:
Similar optimizations for reversed and contains

klaussfreire committed:
Optimize mapped_object.unpack_from

Avoid reliance on struct.unpack_from for the typecode, which
is trivial to do without struct

klaussfreire committed:
Avoid reshape for one-dimensional arrays

klaussfreire committed:
Optimize NDArray representation

Avoid packing the shape when it's implicit in the data buffer

klaussfreire committed:
Several micro-optimizations

- Offsets are Py_ssize_t now. Should not be a problem since that's
  usually a 64-bit signed int, more than enough room for any sane
  buffer offset, and it fits into a native python int which is
  faster to instantiate when necessary. This avoids python 2 longs,
  which are slower, when talking offsets to python code.
- Avoid class attributes when not necessary. Specifically, move
  all header packers to private global constants, to avoid at least
  one attribute access when using them.
- Try to use native arithmetic when manipulating offsets, avoiding
  some python overhead when cythonized.
